File: sdks/arduino/MaestraClient/src/MaestraDiscovery.cpp
/**
 * Maestra Discovery Implementation
 * ESP32-only: mDNS browsing + HTTP provisioning
 */

#include "MaestraDiscovery.h"

#ifdef ESP32

MaestraDiscovery::MaestraDiscovery()
    : _mqttPort(1883) {
    _apiUrl[0] = '\0';
    _mqttBroker[0] = '\0';
    _natsUrl[0] = '\0';
    _wsUrl[0] = '\0';
    _deviceId[0] = '\0';
    _entityId[0] = '\0';
}

// ============================================================================
// mDNS Discovery
// ============================================================================

bool MaestraDiscovery::discoverMaestra(unsigned long timeoutMs) {
    if (!MDNS.begin("maestra-device")) {
        Serial.println("[MaestraDiscovery] mDNS init failed");
        return false;
    }

    Serial.println("[MaestraDiscovery] Browsing for _maestra._tcp ...");

    unsigned long start = millis();
    int found = 0;

    while (millis() - start < timeoutMs) {
        found = MDNS.queryService("maestra", "tcp");
        if (found > 0) break;
        delay(500);
    }

    if (found == 0) {
        Serial.println("[MaestraDiscovery] No Maestra service found");
        return false;
    }

    // Use the first service instance
    IPAddress ip = MDNS.IP(0);
    int port = MDNS.port(0);

    // Build default API URL from host IP and port
    snprintf(_apiUrl, sizeof(_apiUrl), "http://%s:%d", ip.toString().c_str(), port);

    // Default MQTT broker to same host
    strncpy(_mqttBroker, ip.toString().c_str(), sizeof(_mqttBroker) - 1);
    _mqttBroker[sizeof(_mqttBroker) - 1] = '\0';
    _mqttPort = 1883;

    // Parse TXT records for specific service URLs
    int numTxt = MDNS.numTxt(0);
    for (int i = 0; i < numTxt; i++) {
        String key = MDNS.txtKey(0, i);
        String val = MDNS.txt(0, i);

        if (key == "api_url") {
            strncpy(_apiUrl, val.c_str(), sizeof(_apiUrl) - 1);
            _apiUrl[sizeof(_apiUrl) - 1] = '\0';
        } else if (key == "mqtt_broker") {
            strncpy(_mqttBroker, val.c_str(), sizeof(_mqttBroker) - 1);
            _mqttBroker[sizeof(_mqttBroker) - 1] = '\0';
        } else if (key == "mqtt_port") {
            _mqttPort = val.toInt();
        } else if (key == "nats_url") {
            strncpy(_natsUrl, val.c_str(), sizeof(_natsUrl) - 1);
            _natsUrl[sizeof(_natsUrl) - 1] = '\0';
        } else if (key == "ws_url") {
            strncpy(_wsUrl, val.c_str(), sizeof(_wsUrl) - 1);
            _wsUrl[sizeof(_wsUrl) - 1] = '\0';
        }
    }

    Serial.print("[MaestraDiscovery] Found Maestra at ");
    Serial.println(_apiUrl);

    return true;
}

// ============================================================================
// Device Advertisement (POST /devices/discover)
// ============================================================================

bool MaestraDiscovery::advertiseDevice(const char* hardwareId, const char* deviceType, const char* name) {
    if (_apiUrl[0] == '\0') {
        Serial.println("[MaestraDiscovery] No API URL - run discoverMaestra() first");
        return false;
    }

    char url[MAESTRA_URL_BUFFER_SIZE + 32];
    snprintf(url, sizeof(url), "%s/devices/discover", _apiUrl);

    // Build JSON body
    StaticJsonDocument<512> doc;
    doc["hardware_id"] = hardwareId;
    doc["device_type"] = deviceType;
    doc["name"] = name ? name : hardwareId;

    char body[512];
    serializeJson(doc, body);

    HTTPClient http;
    http.begin(url);
    http.addHeader("Content-Type", "application/json");

    int httpCode = http.POST(body);

    if (httpCode != 200 && httpCode != 201) {
        Serial.print("[MaestraDiscovery] advertiseDevice failed, HTTP ");
        Serial.println(httpCode);
        http.end();
        return false;
    }

    // Parse response to get device ID
    String response = http.getString();
    http.end();

    StaticJsonDocument<1024> respDoc;
    DeserializationError err = deserializeJson(respDoc, response);
    if (err) {
        Serial.println("[MaestraDiscovery] Failed to parse advertise response");
        return false;
    }

    const char* id = respDoc["id"] | "";
    strncpy(_deviceId, id, sizeof(_deviceId) - 1);
    _deviceId[sizeof(_deviceId) - 1] = '\0';

    Serial.print("[MaestraDiscovery] Device registered, id=");
    Serial.println(_deviceId);

    return true;
}

// ============================================================================
// Provisioning Poll (GET /devices/{id}/provision)
// ============================================================================

bool MaestraDiscovery::waitForProvisioning(const char* deviceId, unsigned long pollIntervalMs, unsigned long timeoutMs) {
    if (_apiUrl[0] == '\0') {
        Serial.println("[MaestraDiscovery] No API URL - run discoverMaestra() first");
        return false;
    }

    char url[MAESTRA_URL_BUFFER_SIZE + 64];
    snprintf(url, sizeof(url), "%s/devices/%s/provision", _apiUrl, deviceId);

    Serial.println("[MaestraDiscovery] Waiting for provisioning approval ...");

    unsigned long start = millis();

    while (millis() - start < timeoutMs) {
        HTTPClient http;
        http.begin(url);

        int httpCode = http.GET();

        if (httpCode == 200) {
            String response = http.getString();
            http.end();

            StaticJsonDocument<1024> doc;
            DeserializationError err = deserializeJson(doc, response);
            if (err) {
                Serial.println("[MaestraDiscovery] Failed to parse provision response");
                return false;
            }

            // Update connection config from provisioning response
            const char* mqttBroker = doc["mqtt_broker"] | "";
            if (mqttBroker[0] != '\0') {
                strncpy(_mqttBroker, mqttBroker, sizeof(_mqttBroker) - 1);
                _mqttBroker[sizeof(_mqttBroker) - 1] = '\0';
            }

            int mqttPort = doc["mqtt_port"] | 0;
            if (mqttPort > 0) {
                _mqttPort = mqttPort;
            }

            const char* apiUrl = doc["api_url"] | "";
            if (apiUrl[0] != '\0') {
                strncpy(_apiUrl, apiUrl, sizeof(_apiUrl) - 1);
                _apiUrl[sizeof(_apiUrl) - 1] = '\0';
            }

            const char* natsUrl = doc["nats_url"] | "";
            if (natsUrl[0] != '\0') {
                strncpy(_natsUrl, natsUrl, sizeof(_natsUrl) - 1);
                _natsUrl[sizeof(_natsUrl) - 1] = '\0';
            }

            const char* wsUrl = doc["ws_url"] | "";
            if (wsUrl[0] != '\0') {
                strncpy(_wsUrl, wsUrl, sizeof(_wsUrl) - 1);
                _wsUrl[sizeof(_wsUrl) - 1] = '\0';
            }

            const char* entityId = doc["entity_id"] | "";
            if (entityId[0] != '\0') {
                strncpy(_entityId, entityId, sizeof(_entityId) - 1);
                _entityId[sizeof(_entityId) - 1] = '\0';
            }

            Serial.println("[MaestraDiscovery] Provisioning complete!");
            Serial.print("  MQTT: ");
            Serial.print(_mqttBroker);
            Serial.print(":");
            Serial.println(_mqttPort);

            return true;
        }

        http.end();

        if (httpCode == 403) {
            // Still pending - keep polling
        } else if (httpCode == 404) {
            Serial.println("[MaestraDiscovery] Device not found - may have been rejected");
            return false;
        } else {
            Serial.print("[MaestraDiscovery] Unexpected HTTP ");
            Serial.println(httpCode);
        }

        delay(pollIntervalMs);
    }

    Serial.println("[MaestraDiscovery] Provisioning timed out");
    return false;
}

#endif // ESP32

File: sdks/unity/Runtime/MaestraDiscovery.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Text;
using UnityEngine;
using UnityEngine.Networking;
using Newtonsoft.Json;

namespace Maestra
{
    // =========================================================================
    // Data Classes
    // =========================================================================

    /// <summary>
    /// Connection configuration returned by the Maestra Fleet Manager.
    /// Contains all endpoints needed to connect to the platform.
    /// </summary>
    [Serializable]
    public class MaestraConnectionConfig
    {
        [JsonProperty("api_url")]
        public string ApiUrl { get; set; }

        [JsonProperty("nats_url")]
        public string NatsUrl { get; set; }

        [JsonProperty("mqtt_broker")]
        public string MqttBroker { get; set; }

        [JsonProperty("mqtt_port")]
        public int MqttPort { get; set; }

        [JsonProperty("ws_url")]
        public string WsUrl { get; set; }
    }

    /// <summary>
    /// Device registration response from POST /devices/discover.
    /// </summary>
    [Serializable]
    public class DeviceRegistration
    {
        [JsonProperty("id")]
        public string Id { get; set; }

        [JsonProperty("name")]
        public string Name { get; set; }

        [JsonProperty("device_type")]
        public string DeviceType { get; set; }

        [JsonProperty("hardware_id")]
        public string HardwareId { get; set; }

        [JsonProperty("firmware_version")]
        public string FirmwareVersion { get; set; }

        [JsonProperty("ip_address")]
        public string IpAddress { get; set; }

        [JsonProperty("status")]
        public string Status { get; set; }

        [JsonProperty("metadata")]
        public Dictionary<string, object> Metadata { get; set; }

        [JsonProperty("created_at")]
        public string CreatedAt { get; set; }

        [JsonProperty("updated_at")]
        public string UpdatedAt { get; set; }
    }

    /// <summary>
    /// Provisioning configuration returned after device approval.
    /// Contains connection details and optional entity binding.
    /// </summary>
    [Serializable]
    public class ProvisionConfig
    {
        [JsonProperty("device_id")]
        public string DeviceId { get; set; }

        [JsonProperty("provision_status")]
        public string ProvisionStatus { get; set; }

        [JsonProperty("api_url")]
        public string ApiUrl { get; set; }

        [JsonProperty("nats_url")]
        public string NatsUrl { get; set; }

        [JsonProperty("mqtt_broker")]
        public string MqttBroker { get; set; }

        [JsonProperty("mqtt_port")]
        public int MqttPort { get; set; }

        [JsonProperty("ws_url")]
        public string WsUrl { get; set; }

        [JsonProperty("entity_id")]
        public string EntityId { get; set; }

        [JsonProperty("env_vars")]
        public Dictionary<string, object> EnvVars { get; set; }

        /// <summary>
        /// Convert provision config to MaestraConnectionConfig
        /// </summary>
        public MaestraConnectionConfig ToConnectionConfig()
        {
            return new MaestraConnectionConfig
            {
                ApiUrl = ApiUrl,
                NatsUrl = NatsUrl,
                MqttBroker = MqttBroker,
                MqttPort = MqttPort,
                WsUrl = WsUrl,
            };
        }
    }

    // =========================================================================
    // Discovery Helper
    // =========================================================================

    /// <summary>
    /// Static helper class for Maestra device discovery and provisioning.
    /// Provides coroutine-based methods for advertising devices, polling
    /// for provisioning approval, and fetching connection configuration.
    /// </summary>
    public static class MaestraDiscovery
    {
        /// <summary>
        /// Fetch the Maestra connection configuration from a known API URL.
        /// Polls the health endpoint to confirm the server is reachable,
        /// then returns a connection config built from the API URL.
        /// </summary>
        /// <param name="apiUrl">Base URL of the Fleet Manager API (e.g., http://192.168.1.10:8080)</param>
        /// <param name="timeout">Maximum time in seconds to wait for a response</param>
        /// <param name="onSuccess">Called with the connection config on success</param>
        /// <param name="onError">Called with an error message on failure</param>
        public static IEnumerator DiscoverMaestra(
            string apiUrl,
            float timeout,
            Action<MaestraConnectionConfig> onSuccess,
            Action<string> onError)
        {
            string url = $"{apiUrl.TrimEnd('/')}/health";
            float elapsed = 0f;
            float pollInterval = 1f;

            while (elapsed < timeout)
            {
                using (UnityWebRequest request = UnityWebRequest.Get(url))
                {
                    request.timeout = Mathf.CeilToInt(Mathf.Min(5f, timeout - elapsed));
                    yield return request.SendWebRequest();

                    if (request.result == UnityWebRequest.Result.Success)
                    {
                        // Server is reachable — build connection config from the API URL
                        var uri = new System.Uri(apiUrl.TrimEnd('/'));
                        string host = uri.Host;

                        var config = new MaestraConnectionConfig
                        {
                            ApiUrl = apiUrl.TrimEnd('/'),
                            NatsUrl = $"nats://{host}:4222",
                            MqttBroker = host,
                            MqttPort = 1883,
                            WsUrl = $"ws://{host}:8765",
                        };

                        Debug.Log($"[Maestra] Discovered server at {apiUrl}");
                        onSuccess?.Invoke(config);
                        yield break;
                    }
                }

                elapsed += pollInterval;
                if (elapsed < timeout)
                {
                    yield return new WaitForSeconds(pollInterval);
                    elapsed += pollInterval;
                }
            }

            string error = $"Timed out discovering Maestra server at {apiUrl} after {timeout}s";
            Debug.LogError($"[Maestra] {error}");
            onError?.Invoke(error);
        }

        /// <summary>
        /// Advertise this device to the Maestra Fleet Manager via POST /devices/discover.
        /// The device will be registered as "pending" until approved by an admin.
        /// </summary>
        /// <param name="apiUrl">Base URL of the Fleet Manager API</param>
        /// <param name="hardwareId">Unique hardware identifier for this device</param>
        /// <param name="deviceType">Device type string (e.g., "unity", "unreal", "arduino")</param>
        /// <param name="name">Human-readable display name for the device</param>
        /// <param name="onSuccess">Called with the device registration on success</param>
        /// <param name="onError">Called with an error message on failure</param>
        public static IEnumerator AdvertiseDevice(
            string apiUrl,
            string hardwareId,
            string deviceType,
            string name,
            Action<DeviceRegistration> onSuccess,
            Action<string> onError)
        {
            string url = $"{apiUrl.TrimEnd('/')}/devices/discover";

            var body = new Dictionary<string, object>
            {
                { "hardware_id", hardwareId },
                { "device_type", deviceType },
                { "name", name },
                { "metadata", new Dictionary<string, object>
                    {
                        { "platform", Application.platform.ToString() },
                        { "unity_version", Application.unityVersion },
                    }
                },
            };

            string json = JsonConvert.SerializeObject(body);
            byte[] bodyRaw = Encoding.UTF8.GetBytes(json);

            using (UnityWebRequest request = new UnityWebRequest(url, "POST"))
            {
                request.uploadHandler = new UploadHandlerRaw(bodyRaw);
                request.downloadHandler = new DownloadHandlerBuffer();
                request.SetRequestHeader("Content-Type", "application/json");

                yield return request.SendWebRequest();

                if (request.result != UnityWebRequest.Result.Success)
                {
                    string error = $"Failed to advertise device: {request.error} (HTTP {request.responseCode})";
                    Debug.LogError($"[Maestra] {error}");
                    onError?.Invoke(error);
                    yield break;
                }

                try
                {
                    DeviceRegistration registration = JsonConvert.DeserializeObject<DeviceRegistration>(
                        request.downloadHandler.text);
                    Debug.Log($"[Maestra] Device advertised: {registration.Id} (status: {registration.Status})");
                    onSuccess?.Invoke(registration);
                }
                catch (Exception e)
                {
                    string error = $"Failed to parse device registration: {e.Message}";
                    Debug.LogError($"[Maestra] {error}");
                    onError?.Invoke(error);
                }
            }
        }

        /// <summary>
        /// Poll GET /devices/{deviceId}/provision until the device is approved and
        /// provisioning config is available. Returns a 403 while still pending.
        /// </summary>
        /// <param name="apiUrl">Base URL of the Fleet Manager API</param>
        /// <param name="deviceId">Device UUID returned from AdvertiseDevice</param>
        /// <param name="pollInterval">Seconds between polling attempts</param>
        /// <param name="timeout">Maximum total time in seconds to wait for provisioning</param>
        /// <param name="onSuccess">Called with the provision config once approved</param>
        /// <param name="onError">Called with an error message on timeout or unrecoverable failure</param>
        public static IEnumerator WaitForProvisioning(
            string apiUrl,
            string deviceId,
            float pollInterval,
            float timeout,
            Action<ProvisionConfig> onSuccess,
            Action<string> onError)
        {
            string url = $"{apiUrl.TrimEnd('/')}/devices/{deviceId}/provision";
            float elapsed = 0f;

            Debug.Log($"[Maestra] Waiting for provisioning approval (device: {deviceId})...");

            while (elapsed < timeout)
            {
                using (UnityWebRequest request = UnityWebRequest.Get(url))
                {
                    request.timeout = Mathf.CeilToInt(Mathf.Min(10f, timeout - elapsed));
                    yield return request.SendWebRequest();

                    if (request.result == UnityWebRequest.Result.Success)
                    {
                        try
                        {
                            ProvisionConfig config = JsonConvert.DeserializeObject<ProvisionConfig>(
                                request.downloadHandler.text);
                            Debug.Log($"[Maestra] Device provisioned (status: {config.ProvisionStatus})");
                            onSuccess?.Invoke(config);
                            yield break;
                        }
                        catch (Exception e)
                        {
                            string error = $"Failed to parse provision config: {e.Message}";
                            Debug.LogError($"[Maestra] {error}");
                            onError?.Invoke(error);
                            yield break;
                        }
                    }

                    // 403 means device is still pending — keep polling
                    if (request.responseCode == 403)
                    {
                        // Still pending, continue polling
                    }
                    else if (request.responseCode == 404)
                    {
                        string error = "No provisioning record found. Device may have been rejected.";
                        Debug.LogError($"[Maestra] {error}");
                        onError?.Invoke(error);
                        yield break;
                    }
                    else if (request.result != UnityWebRequest.Result.Success)
                    {
                        // Network error or unexpected status — log but keep trying
                        Debug.LogWarning($"[Maestra] Provision poll error: {request.error} (HTTP {request.responseCode})");
                    }
                }

                yield return new WaitForSeconds(pollInterval);
                elapsed += pollInterval;
            }

            string timeoutError = $"Provisioning timed out after {timeout}s (device: {deviceId})";
            Debug.LogError($"[Maestra] {timeoutError}");
            onError?.Invoke(timeoutError);
        }
    }
}
